### PR TITLE
Gear vision

### PR DESCRIPTION
**Pull request by lydiaxing**

### CODE REVIEW

File: arduino/vision_runner/vision_runner.ino [hunk 1]
```diff
@@ -15,38 +15,43 @@ double FOV_X = 75; //horizontal Field of View in degrees
...
 void setup() {
   Serial.begin(9600);
   pixy.init();
 }
 
 void loop() { 
-  blocks = pixy.getBlocks();
+  blocks = pixy.getBlocks(2);
+
+  if (blocks == 2) { //make sure we see two blocks
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Move this comment above the line

File: arduino/vision_runner/vision_runner.ino [hunk 2]
```diff
@@ -15,38 +15,43 @@ double FOV_X = 75; //horizontal Field of View in degrees
...
+     if(abs(pixy.blocks[0].y - pixy.blocks[1].y) > abs(pixy.blocks[0].x - pixy.blocks[1].x)){ //checks that vertical distance between blocks is greater than horizontal distance between blocks, meaning that we are seeing the boiler target
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Move this above the line

File: arduino/vision_runner/vision_runner.ino [hunk 3]
```diff
@@ -4,9 +4,9 @@
 Pixy pixy;
 uint16_t blocks;
 String toSend;
+String TARGET_TYPE;
 char buf[32]; 
-double arr[2];
-
+String STATE;
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Replace with enum http://playground.arduino.cc/Code/Enum

File: arduino/vision_runner/vision_runner.ino [hunk 4]
```diff
@@ -15,38 +15,43 @@ double FOV_X = 75; //horizontal Field of View in degrees
 double FOV_Y = 47; //vertical Field of View in degrees
 double RESOLUTION_WIDTH = 320; //in pixels, 320 x 200
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: If these are constants, make them `const` https://www.arduino.cc/en/Reference/Const (it's like final)

File: arduino/vision_runner/vision_runner.ino (hunk 1 above)
Comment (MiloDavis) [reply arduino/vision_runner/vision_runner.ino]: What if you can see both a gear and the boiler?

File: arduino/vision_runner/vision_runner.ino [hunk 5]
```diff
@@ -15,38 +15,43 @@ double FOV_X = 75; //horizontal Field of View in degrees
...
 }
 
 void loop() { 
-  blocks = pixy.getBlocks();
+  blocks = pixy.getBlocks(2);
+
+  if (blocks == 2) { //make sure we see two blocks
+     if(abs(pixy.blocks[0].y - pixy.blocks[1].y) > abs(pixy.blocks[0].x - pixy.blocks[1].x)){ //checks that vertical distance between blocks is greater than horizontal distance between blocks, meaning that we are seeing the boiler target
+       STATE = "BOIL";
+       updateValuesBoiler();
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Don't have this operate by side effect.  Have this take parameters and return the values

File: arduino/vision_runner/vision_runner.ino [hunk 6]
```diff
@@ -15,38 +15,43 @@ double FOV_X = 75; //horizontal Field of View in degrees
...
+       writeBytes(distanceToTarget, angleToTarget_x);
+     }else{ //then that means we're seeing the gears!
+       //TODO: CHANGE THESE CONSTANTS
+       VERTICAL_ZERO_OFFSET = 23.5;
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Don't assign to these here

File: arduino/vision_runner/vision_runner.ino [hunk 7]
```diff
@@ -15,38 +15,43 @@ double FOV_X = 75; //horizontal Field of View in degrees
...
- 
-     toSend = String(getDistance(), 3).substring(0,4) + ":" + String(angleToTarget_x, 3).substring(0,4) + "\n";
-     byte sendBytes[toSend.length() + 1];
-     toSend.getBytes(sendBytes, toSend.length() + 1);
-     Serial.write(sendBytes, toSend.length() + 1);
-     Serial.flush();
-  }  
+  //If we don't see anything, just send -1000 for both values!
+  STATE = "NONE";
+  writeBytes(-1000.0, -1000.0);
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Can you not write any value here?

File: arduino/vision_runner/vision_runner.ino [hunk 8]
```diff
@@ -66,3 +71,23 @@ double degreesToRadians(double deg){
...
 }
+
+void updateValuesBoiler(){
+  angleToTarget_x = getHorizontalAngleOffset(pixy.blocks[0].x);
+  angleToTarget_y = getVerticalAngleOffset(pixy.blocks[0].y);
+  distanceToTarget = getDistance();
+}
+
+void updateValuesGears(){
+  angleToTarget_x = getHorizontalAngleOffset((pixy.blocks[0].x + pixy.blocks[1].x) / 2); //find the midline between the gear targets, this is what we want to turn to
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Move comment above line

File: arduino/vision_runner/vision_runner.ino [hunk 9]
```diff
@@ -66,3 +71,23 @@ double degreesToRadians(double deg){
...
+  angleToTarget_y = getVerticalAngleOffset(pixy.blocks[0].y);
+  distanceToTarget = getDistance();
+}
+
+void writeBytes(double distance, double angle){
```
Comment (MiloDavis) [inline arduino/vision_runner/vision_runner.ino]: Investigate writing two doubles.   I still think that's how this should be done

File: src/com/nutrons/steamworks/Feeder.java [hunk 10]
```diff
@@ -2,17 +2,19 @@
 
 import com.nutrons.framework.Subsystem;
 import com.nutrons.framework.controllers.LoopSpeedController;
+import com.nutrons.framework.controllers.RunAtPowerEvent;
+import com.nutrons.framework.controllers.Talon;
 
 public class Feeder implements Subsystem {
 
-  private final LoopSpeedController intakeController;
+  private final Talon intakeMotor;
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Feeder.java]: We know from the type its a motor.

File: src/com/nutrons/steamworks/Feeder.java [hunk 11]
```diff
@@ -2,17 +2,19 @@
...
-  public Feeder(LoopSpeedController intakeController) {
-    this.intakeController = intakeController;
+  public Feeder(Talon intakeMotor) {
+    this.intakeMotor = intakeMotor;
   }
 
   @Override
   public void registerSubscriptions() {
-    intakeController.runAtPower(1.0);
+    new RunAtPowerEvent(1.0).actOn(intakeMotor);
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Feeder.java]: Move 1.0 up to the top of the class as a constant

File: src/com/nutrons/steamworks/Hopper.java [hunk 12]
```diff
@@ -2,19 +2,21 @@
...
 import com.nutrons.framework.controllers.LoopSpeedController;
+import com.nutrons.framework.controllers.RunAtPowerEvent;
+import com.nutrons.framework.controllers.Talon;
 
 public class Hopper implements Subsystem {
 
   // TODO: tune as needed
   private static final double SPIN_POWER = 0.9;
-  private final LoopSpeedController hopperController;
+  private final Talon hopperMotor;
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Hopper.java]: We also know this is a motor

File: src/com/nutrons/steamworks/Hopper.java [hunk 13]
```diff
@@ -2,19 +2,21 @@
...
-  public Hopper(LoopSpeedController hopperController) {
-    this.hopperController = hopperController;
+  public Hopper(Talon hopperMotor) {
+    this.hopperMotor = hopperMotor;
   }
 
   @Override
   public void registerSubscriptions() {
-    hopperController.runAtPower(SPIN_POWER);
+    new RunAtPowerEvent(1.0).actOn(hopperMotor);
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Hopper.java]: This should also be moved up to be a constant

File: src/com/nutrons/steamworks/Shooter.java [hunk 14]
```diff
@@ -14,7 +17,7 @@ public Shooter(LoopSpeedController shooterController) {
 
   @Override
   public void registerSubscriptions() {
-    shooterController.setPID(0.05, 0.0, 0.33, 0.035);
-    shooterController.setSetpoint(2950.0);
+    Flowable<ControllerEvent> source = toFlow(() -> new LoopPropertiesEvent(2950.0, 0.05, 0.0, 0.33, 0.035));
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Shooter.java]: Use Flowable.just to simplify this

File: src/com/nutrons/steamworks/Shooter.java [hunk 15]
```diff
@@ -14,7 +17,7 @@ public Shooter(LoopSpeedController shooterController) {
...
+        source.mergeWith(toFlow(() -> new LoopModeEvent(ControlMode.LOOP_SPEED))).subscribe(shooterController);
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Shooter.java]: and this

Comment (MiloDavis) [reply src/com/nutrons/steamworks/Shooter.java]: Also @totallyashes is adding imperative methods which might be the better choice here.

File: src/com/nutrons/steamworks/Vision.java [hunk 16]
```diff
@@ -5,21 +5,26 @@
...
     this.dataStream = dataStream;
-    this.dataStreamDouble = this.dataStream
-        .filter(x -> x.length == 10)
+    this.dataStreamString = this.dataStream
+        .filter(x -> x.length == 17)
         .map(x -> new String(x, "UTF-8"))
-        .map(x -> x.split(":")).filter(x -> x.length == 2)
-        .map(x -> new Double[]{Double.valueOf(x[0]),
-            Double.valueOf(x[1])}); //Returns a double array[distance, angle]
+        .map(x -> x == DUMMY_VALUE ? "NONE:0.0:0.0" : x) //vision will change any dummy values to 0.0
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Vision.java]: Better to filter them

File: src/com/nutrons/steamworks/Vision.java [hunk 17]
```diff
@@ -5,21 +5,26 @@
...
   private Flowable<Double> distance;
+  private Flowable<String> state;
+  private static final String DUMMY_VALUE = "NONE:-1000:-1000"; //Arduino sends -1000.0 over serial when it doesn't see anything, to prevent
+                                                               //robot sending an exception "no serial port found"
 
   Vision(Flowable<byte[]> dataStream) {
     this.dataStream = dataStream;
-    this.dataStreamDouble = this.dataStream
-        .filter(x -> x.length == 10)
+    this.dataStreamString = this.dataStream
```
Comment (MiloDavis) [inline src/com/nutrons/steamworks/Vision.java]: Once again, consider using doubles encoded as bytes

**Review by MiloDavis — DISMISSED** [review]

File: arduino/vision_runner/vision_runner.ino (hunk 4 above)
Comment (lydiaxing) [reply arduino/vision_runner/vision_runner.ino]: struct

File: arduino/vision_runner/vision_runner.ino (hunk 5 above)
Comment (lydiaxing) [reply arduino/vision_runner/vision_runner.ino]: ex getDistnace(param1, param2);

File: arduino/vision_runner/vision_runner.ino (hunk 7 above)
Comment (lydiaxing) [reply arduino/vision_runner/vision_runner.ino]: write a ping to arduino
ask if arduino is real

File: arduino/vision_runner/vision_runner.ino (hunk 1 above)
Comment (lydiaxing) [reply arduino/vision_runner/vision_runner.ino]: Will know once mechanical decides where to mount camera

File: arduino/vision_runner/vision_runner.ino (hunk 7 above)
Comment (lydiaxing) [reply arduino/vision_runner/vision_runner.ino]: I think this is lower priority, i'll  put it on scrum as a todo

File: src/com/nutrons/steamworks/Shooter.java (hunk 14 above)
Comment (lydiaxing) [reply src/com/nutrons/steamworks/Shooter.java]: this is fixed on another PR

File: src/com/nutrons/steamworks/Shooter.java (hunk 15 above)
Comment (lydiaxing) [reply src/com/nutrons/steamworks/Shooter.java]: also fixed on another PR

**Review by kreeve — APPROVED** [review]

**Pull request merged**